p22/src/song.rs

```diff
@@ -0,0 +1,236 @@
+static SONG_PARTS: [&str; 12] = [
+    "A partridge in a pear tree",
+    "Two turtle doves",
+    "Three French hens",
+    "Four calling birds",
+    "Five gold rings",
+    "Six geese a-laying",
+    "Seven swans a-swimming",
+    "Eight maids a-milking",
+    "Nine ladies dancing",
+    "Ten lords a-leaping",
+    "Eleven pipers piping",
+    "Twelve drummers drumming",
+];
+
+static SONG_DAYS: [&str; 12] = [
+    "first", "second", "third", "fourth", "fifth", "sixth", "seventh", "eighth", "ninth", "tenth",
+    "eleventh", "twelfth",
+];
+
+pub struct Song {}
+
+impl Default for Song {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Song {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    pub fn compute_all(&self) -> Vec<String> {
+        let mut result: Vec<String> = Vec::with_capacity(12);
+        for day in 0..SONG_DAYS.len() {
+            result.push(self.compute(day))
+        }
+        assert_eq!(result.len(), SONG_DAYS.len());
+        result
+    }
+
+    fn nth_day(&self, day: usize) -> String {
+        let ordinal = SONG_DAYS[day];
+
+        format!(
+            "On the {} day of Christmas, my true love sent to me",
+            ordinal
+        )
+        .to_string()
+    }
+
+    pub fn compute(&self, day: usize) -> String {
+        let terminator = if day == 11 { "!" } else { "." };
+        format!("{} {}{}", self.nth_day(day), self.rest(day), terminator)
+    }
+
+    // https://stackoverflow.com/questions/38406793/why-is-capitalizing-the-first-letter-of-a-string-so-convoluted-in-rust
+    fn make_first_lowercase(&self, string: &mut str) {
+        if let Some(r) = string.get_mut(0..1) {
+            r.make_ascii_lowercase();
+        }
+    }
+    fn rest(&self, day: usize) -> String {
+        if day == 0 {
+            return SONG_PARTS[day].to_string();
+        }
+        let selected_days = 1..day + 1;
+
+        let initial: String =
+            // https://doc.rust-lang.org/beta/alloc/string/struct.String.html#method.with_capacity
+            // s.push(char)
+            // s.push(&String)
+            format!(
+                "{},",
```

Comment:
You could've just used `push(',')` on the `join`ed `String`.

Reply:
good catch - it's cleaner